[clang][Sema] Bugfix for choosing the more specialized overload

HoBoIs opened the pull request:
There was a bug in Clang where it couldn't choose which overload candidate was more specialized if it was comparing a member-function to a non-member function. Previously, this was detected as an ambiguity, now Clang chooses correctly.

This patch fixes the bug by fully implementing CWG2445 and moving the template transformation described in `[temp.func.order]` paragraph 3 from `isAtLeastAsSpecializedAs()` to `Sema::getMoreSpecializedTemplate()` so we have the transformed parameter list during the whole comparison. Also, to be able to add the correct type for the implicit object parameter `Sema::getMoreSpecializedTemplate()` has new parameters for the object type.

Fixes #74494, fixes #82509

HoBoIs committed:
Bugfix for choosing the more specialized overload

There was a bug in clang where it couldn't choose which overload candidate is
more specialized if it was comparing a member-function to a non-member
function. Previously, this was detected as an ambigouity, now clang chooses correctly.

This patch fixes the bug by fully implementing CWG2445 and moving the template
transformation described in [temp.func.order] paragraph 3 from
isAtLeastAsSpecializedAs to Sema::getMoreSpecializedTemplate so we have the
transformed parameter list during the whole comperrassion. Also, to be able
to add the correct type for the implicit object parameter
Sema::getMoreSpecializedTemplate has new parameters for the object type.

HoBoIs commented:
@zygoloid @erichkeane Could you take a look?

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5333,11 +5333,31 @@ bool Sema::CheckIfFunctionSpecializationIsImmediate(FunctionDecl *FD,
 
+static QualType GetImplicitObjectParameterTypeCXX20(ASTContext &Context,
+                                                    const CXXMethodDecl *Method,
+                                                    QualType rawType,
```
Coding standard says parms are capital first letter.

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5333,11 +5333,31 @@ bool Sema::CheckIfFunctionSpecializationIsImmediate(FunctionDecl *FD,
   return false;
 }
 
+static QualType GetImplicitObjectParameterTypeCXX20(ASTContext &Context,
```
I don't think splitting this method up into 2 like this is necessary, I think these two should be the same function with the language-based split for the condition of the 'if' statement.

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5548,13 +5514,113 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+    const CXXMethodDecl *Method1 = dyn_cast<CXXMethodDecl>(FD1);
+    const CXXMethodDecl *Method2 = dyn_cast<CXXMethodDecl>(FD2);
+
+    if (getLangOpts().CPlusPlus20) {
```
Same here, these two branches are REALLY similar and should probably be combined.

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5548,13 +5514,113 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+      shouldConvert2 = Method2 && !Method2->isExplicitObjectMemberFunction();
+      if (shouldConvert1) {
+        bool isR2 =
+            Method2 && !Method2->isExplicitObjectMemberFunction()
```
This whole line is `shouldConvert2`.

erichkeane reviewed (commented):
This needs tests plus a release note.  

Also, there is a lot of repeated code between the pre-c++20 and c++20 versions that will make this really hard to maintain, particularly with core issues, that I'd like to see merged better.

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5548,13 +5504,100 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+    if (shouldConvert1) {
+      bool isR2 =
+          getLangOpts().CPlusPlus20 &&
+          (shouldConvert1
```
Think this line is a copy/paste error?  We already know the value of shouldConvert1 (and 2 below).

HoBoIs replied:
Actually I just mixed up `shouldConvert1` and `shouldConvert2`

HoBoIs replied:
fixed

erichkeane commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5548,13 +5504,100 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+    // C++ [temp.func.order]p5:
+    //   The presence of unused ellipsis and default arguments has no effect on
+    //   the partial ordering of function templates.
+    if (Args1.size() > NumComparedArguments)
```
```suggestion
    Args1.resize(std::max(Args1.size(), NumComparedArguments));
    Args2.resize(std::max(Args2.size(), NumComparedArguments));
```

HoBoIs replied:
`Args1.resize(std::max(Args1.size(), NumComparedArguments));` is not functionally equivalent to what I wrote, and will not work as intended. I need to drop elements, not to add new default constructed ones. I think you meant to write `min` instead of `max`. 
 Still I don't see why `Args1.resize(min(Args1.size(), NumComparedArguments));` is preferable to the original.

erichkeane replied:
Seemingly the original is more confusing as to what it is doing than I thought :)    I thought that the 'if' statements were overly verbose for something pretty trivial and the 1 liners made this less distracting along with reading the rest of the function.

HoBoIs replied:
How can I make it less confusing? Should I put in a comment explaining that in `Args1` and `Args2` after the first `NumComparedArguments` elements there can only be unused ellipsis and default arguments? (It already states that  unused ellipsis and default arguments has no effect on the partial ordering of function templates. Just above the if)

erichkeane replied:
I thought the `resize` model was pretty clear, but that is a pattern I use frequently so I thought it read easily.

HoBoIs replied:
Done

erichkeane reviewed (commented):
This is close I think, just needs tests + ReleaseNotes.rst entry.

Endilll commented on clang/www/cxx_dr_status.html:
```diff
@@ -14478,7 +14478,7 @@ <h2 id="cxxdr">C++ defect report implementation status</h2>
     <td>C++20</td>
     <td>Partial ordering with rewritten candidates</td>
-    <td class="unknown" align="center">Unknown</td>
+    <td class="unreleased" align="center">Clang 18</td>
```
You shouldn't edit this file by hand. Instead, you should place a test in `/clang/test/CXX/drs/24xx.cpp` with a `// dr2445: 18` status, then run `/clang/www/make_cxx_dr_status` script. Follow the example of other tests in `24xx.cpp` while doing that.

HoBoIs replied:
I'm on it. While we are at here: should it be `18` or `19` ? As I can see 18 is almost out.

Endilll replied:
`19`, unless you're planning to cherry-pick this into 18 branch (I don't advise to do that this late into pre-release phase).

HoBoIs replied:
done

Endilll commented on clang/test/CXX/drs/dr24xx.cpp:
```diff
@@ -68,3 +68,35 @@ template<A> struct X {};
 X<1> x;
 #endif
 }
+namespace dr2445 { // dr2445: 19
```
It would be nice to have an empty line before the new namespace.

Endilll reviewed (approved):
DR changes looks good now, save for a nitpick.
However, I'm leaving it to other reviewers to check contents of the test. It looks like an example from the DR, but I don't feel qualified to say whether this test has sufficient coverage.

HoBoIs commented:
> DR changes looks good now, save for a nitpick. However, I'm leaving it to other reviewers to check contents of the test. It looks like an example from the DR, but I don't feel qualified to say whether this test has sufficient coverage.

I have added tests to other files what are relevant for the DR too. Should I move them here?

Endilll commented:
> I have added tests to other files what are relevant for the DR too. Should I move them here?

I trust your judgement here. Your PR touches a topic complicated enough that I don't feel qualified to make the judgement myself.

My stance is that DR should be considered narrow in scope. So if you think that your additional tests are directly related to what the DR is about, go ahead. For instance, when DR is about cv-qualifiers in variable declarations, but the provided examples only cover `const`, I consider additional test for `volatile` to be directly related. On the other hand, if during implementation you realize that something is going on with cv-qualifiers on non-static member functions, and you want to test that, then it's a related issue, but not a part of the DR.

In my view, DR test suite prioritizes being on spot more than just containing sheer volume of tests.

HoBoIs commented:
I have added tests for rvalref-s and `volatile`-s.

HoBoIs commented:
@erichkeane a question: this PR is partly a fix for a DR for c++20. Right now I only apply the fix for C++20 and newer. Is there a reason for not applying this fix for older standards? I don't see how it would break any existing code, and making the fix retroactive doesn't make any test fail.

HoBoIs commented:
> I don't think checking for c++20 is useful because rewritten candidates are a c++20- feature anyway

Yes, but the resolution of the DR also talks about the reference type of the object parameter. (If I apply the patch retroactively the code will get simpler.)

HoBoIs commented:
@erichkeane Thank you for the suggestions, I have removed the checking for C++20 and I also changed to the `Args.resize(std::min(...))` based on what you suggested. Is there anything else to be done?

erichkeane reviewed (approved):
Only some minor things left, so approving.  Mostly, for the 'type' and 'context', just do them the way they would read:
So "Object One Type" (plus we use Ty to mean type).  Same with Context.

Finally, `SmallVector` doesn't need a size set, so just using the 'default' value (which it calculates for you!)  is sufficient here.

Feel free to merge when CI agrees with you on the above/below changes.

whisperity commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5333,38 +5333,37 @@ bool Sema::CheckIfFunctionSpecializationIsImmediate(FunctionDecl *FD,
+                                               const CXXMethodDecl *Method,
+                                               QualType RawType,
+                                               bool IsOtherRvr) {
+  // C++20 [temp.func.order]p3.1, p3.2:
+  //- The type X(M ) is “rvalue reference to cv A” if the optional ref-qualifier
+  //  of M is && or if M has no ref-qualifier and the positionally-corresponding
+  //  parameter of the other transformed template has rvalue reference type;
+  //  if this determination depends recursively upon whether X(M ) is an rvalue
+  //  reference type, it is not considered to have rvalue reference type.
+  //- Otherwise, X(M ) is “lvalue reference to cv A”.
```
```suggestion
  // C++20 [temp.func.order]p3.1, p3.2:
  //  - The type X(M) is "rvalue reference to cv A" if the optional
  //    ref-qualifier of M is && or if M has no ref-qualifier and the
  //    positionally-corresponding parameter of the other transformed
  //    template has rvalue reference type; if this determination depends
  //    recursively upon whether X(M) is an rvalue reference type, it
  //    is not considered to have rvalue reference type.
  //
  //  - Otherwise, X(M) is "lvalue reference to cv A".
```

whisperity commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5548,13 +5493,76 @@ static bool isAtLeastAsSpecializedAs(Sema &S,
+    // "that is a member function with no expicit object argument".
+    // Otherwise the ordering rules for methods with expicit objet arguments
+    // against anything else make no sense.
+    shouldConvert1 = Method1 && !Method1->isExplicitObjectMemberFunction();
+    shouldConvert2 = Method2 && !Method2->isExplicitObjectMemberFunction();
```
```suggestion
    ShouldConvert1 = Method1 && !Method1->isExplicitObjectMemberFunction();
    ShouldConvert2 = Method2 && !Method2->isExplicitObjectMemberFunction();
```

Style nit formatting, if we're changing the code even though it might largely be a move/refactor, you can fix the formatting now!

whisperity commented on clang/lib/Sema/SemaTemplateDeduction.cpp:
```diff
@@ -5568,12 +5576,28 @@ FunctionTemplateDecl *Sema::getMoreSpecializedTemplate(
-  bool Variadic1 = NumParams1 && FD1->parameters().back()->isParameterPack();
-  bool Variadic2 = NumParams2 && FD2->parameters().back()->isParameterPack();
+
+  SmallVector<QualType> param1;
```
```suggestion
  SmallVector<QualType> Param1;
```

Ditto, everywhere!

whisperity reviewed (changes requested):
(Only style nits!)

HoBoIs committed:
Format fixes

whisperity reviewed (approved):
LGTM.
